From f5eb8011c74c79f8c41608e2b82d8c37beb11002 Mon Sep 17 00:00:00 2001
From: Yixun Lan <dlan@gentoo.org>
Date: Fri, 10 Apr 2026 10:46:39 +0800
Subject: [PATCH 04/33] clock: sunxi: add rtc ccu support for A733

The rtc ccu modules contains various clock settings for SERDES, HDMI, UFS.

Signed-off-by: Yixun Lan <dlan@gentoo.org>
---
 drivers/clk/sunxi/Kconfig                   |  7 ++++++
 drivers/clk/sunxi/Makefile                  |  1 +
 drivers/clk/sunxi/clk_a733_rtc.c            | 24 +++++++++++++++++++++
 drivers/clk/sunxi/clk_sunxi.c               |  5 +++++
 include/dt-bindings/clock/sun60i-a733-rtc.h | 16 ++++++++++++++
 5 files changed, 53 insertions(+)
 create mode 100644 drivers/clk/sunxi/clk_a733_rtc.c
 create mode 100644 include/dt-bindings/clock/sun60i-a733-rtc.h

diff --git a/drivers/clk/sunxi/Kconfig b/drivers/clk/sunxi/Kconfig
index 03133ba86df..3fd57ab453e 100644
--- a/drivers/clk/sunxi/Kconfig
+++ b/drivers/clk/sunxi/Kconfig
@@ -157,4 +157,11 @@ config CLK_SUN60I_A733_R
 	  This enables common clock driver support for the PRCM
 	  in Allwinner A733 SoCs.
 
+config CLK_SUN60I_A733_RTC
+	bool "Clock driver for Allwinner A733 RTC CCU"
+	default MACH_SUN60I_A733
+	help
+	  This enables common clock driver support for the RTC CCU
+	  in Allwinner A733/T736 SoCs.
+
 endif # CLK_SUNXI
diff --git a/drivers/clk/sunxi/Makefile b/drivers/clk/sunxi/Makefile
index 943e47612ce..f769335efe6 100644
--- a/drivers/clk/sunxi/Makefile
+++ b/drivers/clk/sunxi/Makefile
@@ -29,3 +29,4 @@ obj-$(CONFIG_CLK_SUN55I_A523) += clk_a523.o
 obj-$(CONFIG_CLK_SUN55I_A523_R) += clk_a523_r.o
 obj-$(CONFIG_CLK_SUN60I_A733) += clk_a733.o
 obj-$(CONFIG_CLK_SUN60I_A733_R) += clk_a733_r.o
+obj-$(CONFIG_CLK_SUN60I_A733_RTC) += clk_a733_rtc.o
diff --git a/drivers/clk/sunxi/clk_a733_rtc.c b/drivers/clk/sunxi/clk_a733_rtc.c
new file mode 100644
index 00000000000..d9f6ed15445
--- /dev/null
+++ b/drivers/clk/sunxi/clk_a733_rtc.c
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+
+#include <clk-uclass.h>
+#include <dm.h>
+#include <clk/sunxi.h>
+#include <dt-bindings/clock/sun60i-a733-rtc.h>
+#include <linux/bitops.h>
+
+static struct ccu_clk_gate a733_rtc_gates[] = {
+	[CLK_IOSC]				= GATE_DUMMY,
+	[CLK_OSC32K]			= GATE_DUMMY,
+	[CLK_HOSC]				= GATE_DUMMY,
+	[CLK_RTC_32K]			= GATE_DUMMY,
+	[CLK_OSC32K_FANOUT]		= GATE_DUMMY,
+	[CLK_HOSC_SERDES1]		= GATE(0x16c, BIT(5)),
+	[CLK_HOSC_SERDES0]		= GATE(0x16c, BIT(4)),
+	[CLK_HOSC_HDMI]			= GATE(0x16c, BIT(1)),
+	[CLK_HOSC_UFS]			= GATE(0x16c, BIT(0)),
+};
+
+const struct ccu_desc a733_rtc_ccu_desc = {
+	.gates = a733_rtc_gates,
+	.num_gates = ARRAY_SIZE(a733_rtc_gates),
+};
diff --git a/drivers/clk/sunxi/clk_sunxi.c b/drivers/clk/sunxi/clk_sunxi.c
index ad90a3744a2..e7f98359017 100644
--- a/drivers/clk/sunxi/clk_sunxi.c
+++ b/drivers/clk/sunxi/clk_sunxi.c
@@ -130,6 +130,7 @@ extern const struct ccu_desc a523_ccu_desc;
 extern const struct ccu_desc a523_r_ccu_desc;
 extern const struct ccu_desc a733_ccu_desc;
 extern const struct ccu_desc a733_r_ccu_desc;
+extern const struct ccu_desc a733_rtc_ccu_desc;
 
 static const struct udevice_id sunxi_clk_ids[] = {
 #ifdef CONFIG_CLK_SUN4I_A10
@@ -243,6 +244,10 @@ static const struct udevice_id sunxi_clk_ids[] = {
 #ifdef CONFIG_CLK_SUN60I_A733_R
 	{ .compatible = "allwinner,sun60i-a733-r-ccu",
 	  .data = (ulong)&a733_r_ccu_desc },
+#endif
+#ifdef CONFIG_CLK_SUN60I_A733_RTC
+	{ .compatible = "allwinner,sun60i-a733-rtc",
+	  .data = (ulong)&a733_rtc_ccu_desc },
 #endif
 	{ }
 };
diff --git a/include/dt-bindings/clock/sun60i-a733-rtc.h b/include/dt-bindings/clock/sun60i-a733-rtc.h
new file mode 100644
index 00000000000..8a2b5facad7
--- /dev/null
+++ b/include/dt-bindings/clock/sun60i-a733-rtc.h
@@ -0,0 +1,16 @@
+/* SPDX-License-Identifier: GPL-2.0-only OR MIT */
+
+#ifndef _DT_BINDINGS_CLK_SUN60I_A733_RTC_H_
+#define _DT_BINDINGS_CLK_SUN60I_A733_RTC_H_
+
+#define CLK_IOSC		0
+#define CLK_OSC32K		1
+#define CLK_HOSC		2
+#define CLK_RTC_32K		3
+#define CLK_OSC32K_FANOUT	4
+#define CLK_HOSC_SERDES1	5
+#define CLK_HOSC_SERDES0	6
+#define CLK_HOSC_HDMI		7
+#define CLK_HOSC_UFS		8
+
+#endif /* _DT_BINDINGS_CLK_SUN60I_A733_RTC_H_ */
-- 
2.53.0

